src/main/webapp/dev/js/main/instructorCourseEnrollPage.js

```diff
@@ -43,39 +69,135 @@ const handsontable = new Handsontable(container, {
+ * @returns {Promise} the state of the result from the AJAX request
+ */
+function getAjaxStudentList(displayIcon) {
+    return new Promise((resolve, reject) => {
+        const $spreadsheetForm = $('#student-spreadsheet-form');
+        $.ajax({
+            type: 'POST',
+            url: '/page/instructorCourseEnrollAjaxPage',
+            cache: false,
+            data: {
+                courseid: $spreadsheetForm.children(`input[name="${ParamsNames.COURSE_ID}"]`).val(),
+                user: $spreadsheetForm.children(`input[name="${ParamsNames.USER_ID}"]`).val(),
+            },
+            beforeSend() {
+                displayIcon.html('<img height="25" width="25" src="/images/ajax-preload.gif">');
+            },
+        })
+                .done(resolve)
+                .fail(reject);
+    });
+}
+
+/**
+ * Handles how the panels are displayed, including rendering the spreadsheet interface.
+ */
+function adjustStudentsPanelView($panelHeading, panelCollapse,
+        displayIcon, toggleChevron) {
+    toggleStudentsPanel($panelHeading, panelCollapse,
+            displayIcon, toggleChevron);
+    dataHandsontable.render(); // needed as the view is buggy after collapsing the panel
+}
+
+/**
+ * Expands "Existing students" panel and loads existing students' data (if spreadsheet is not empty)
+ * into the spreadsheet interface. Spreadsheet interface would be shown after expansion.
+ * The panel will be collapsed otherwise if the spreadsheet interface is already shown.
+ */
+function expandCollapseExistingStudentsPanel() {
+    const $panelHeading = $(this);
+    const panelCollapse = $panelHeading.parent().children('.panel-collapse');
+    const displayIcon = $panelHeading.children('.display-icon');
+    const toggleChevron = $panelHeading.parent().find('.glyphicon-chevron-down, .glyphicon-chevron-up');
+
+    // perform AJAX only if existing students' spreadsheet is empty
+    if (getSpreadsheetLength(dataHandsontable.getData()) === 0) {
+        getAjaxStudentList(displayIcon)
+                .then((data) => {
+                    if (data.students.length === 0) {
+                        displayNoExistingStudents(displayIcon);
+                    } else {
+                        loadExistingStudentsData(data.students);
+                        adjustStudentsPanelView($panelHeading, panelCollapse,
+                                displayIcon, toggleChevron);
+                    }
+                }).catch(() => {
+                    displayErrorExecutingAjax(displayIcon);
+                });
+    } else {
+        adjustStudentsPanelView($panelHeading, panelCollapse,
+                displayIcon, toggleChevron);
+    }
+}
+
+/**
+ * Expands "New students" panel. Spreadsheet interface would be shown after expansion, including its affiliated buttons.
+ * The panel will be collapsed otherwise if the spreadsheet interface is already shown.
+ */
+function expandCollapseNewStudentsPanel() {
```

Comment:
Don't you think this function is very similar to the above? You can make your code more concise by grouping these 2 functions together like this:
```
function makePanelClickHandler(isExistingStudentsPanel) {
    return () => {
        // common code
        if (isExistingStudentsPanel) {
            // existing student panel only code
        }
    }
}
```

Reply:
An immediate suspect is `$(this) `. `this` may not behave as you would expect in anonymous function.